sonic_platform_base/sonic_xcvr/api/public/cmisCDB.py

```diff
@@ -0,0 +1,430 @@
+            if (epl_len % PAGE_LENGTH) != 0:
+                pages += 1
+            # write to page 0xA0 - 0xAF (max of 16 pages)
+            for pageoffset in range(pages):
+                next_page = 0xa0 + pageoffset
+                if PAGE_LENGTH*(pageoffset + 1) <= epl_len:
+                    datachunk = data[PAGE_LENGTH*pageoffset : PAGE_LENGTH*(pageoffset + 1)]
+                    self.xcvr_eeprom.write_raw(next_page*PAGE_LENGTH+INIT_OFFSET, PAGE_LENGTH, datachunk)
+                else:
+                    datachunk = data[INIT_OFFSET*pageoffset : ]
+                    self.xcvr_eeprom.write_raw(next_page*PAGE_LENGTH+INIT_OFFSET, len(datachunk), datachunk)
+        else:
+            sections = epl_len // writelength
+            if (epl_len % writelength) != 0:
+                sections += 1
+            # write to page 0xA0 - 0xAF (max of 16 pages), with length of writelength per piece
+            for offset in range(0, epl_len, writelength):
+                if offset + writelength <= epl_len:
+                    datachunk = data[offset : offset + writelength]
+                    self.xcvr_eeprom.write_raw(0xA0*PAGE_LENGTH+offset+INIT_OFFSET, writelength, datachunk)
+                else:
+                    datachunk = data[offset : ]
+                    self.xcvr_eeprom.write_raw(0xA0*PAGE_LENGTH+offset+INIT_OFFSET, len(datachunk), datachunk)
+        subtimeint = time.time()-subtime
+        logger.info('%dB write time:  %.2fs' %(epl_len, subtimeint))
+        cmd = bytearray(b'\x01\x04\x08\x00\x04\x00\x00\x00')
+        addr_byte = struct.pack('>L',addr)
+        cmd += addr_byte
+        cmd[133-INIT_OFFSET] = self.cdb_chkcode(cmd)
+        for attemp in range(MAX_TRY):
+            self.write_cdb(cmd)
+            time.sleep(2)
+            status = self.cdb1_chkstatus()
+            if (status != 0x1):
+                logger.info('CDB1 status: Fail. CDB1 status %d' %status)
+                continue
+            else:
+                break
+        return status
+
+    # FW download complete
+    def validate_fw_image(self):
+        '''
+        When this command is issued, the module shall validate the complete
+        image and then return success or failure
+        '''
+        cmd = bytearray(b'\x01\x07\x00\x00\x00\x00\x00\x00')
+        cmd[133-INIT_OFFSET] = self.cdb_chkcode(cmd)
+        for attemp in range(MAX_TRY):
+            self.write_cdb(cmd)
+            time.sleep(2)
+            status = self.cdb1_chkstatus()
+            if (status != 0x1):
+                logger.info('CDB1 status: Fail. CDB1 status %d' %status)
+                continue
+            else:
+                break
+        return status
+
+    # Run FW image
+    # mode:
+    # 00h = Traffic affecting Reset to Inactive Image.
+    # 01h = Attempt Hitless Reset to Inactive Image
+    # 02h = Traffic affecting Reset to Running Image.
+    # 03h = Attempt Hitless Reset to Running Image
+    def run_fw_image(self, mode = 0x01):
```

Comment:
will all module support mode=1? if not we don't seem to handle gracefully?

Reply:
Mode 1 is  "Attempt Hitless Reset to Inactive Image". It won't break even if it has to be hitless.